### PR TITLE
Client-prediction tracking ("Client-mapped entities")

### PR DESCRIPTION
**Pull request by RJ**
This is working fine, in the test and in my game (which has a misprediction cleanup system similar to the new docs describe).

I'm currently sending Entity::PLACEHOLDER where there is no prediction, so still need to decide on the packet format.
EDIT: pushed a change to encode compactly now

With a bit more code complexity, like a shared config between client and server, we can make it totally opt-in and zero cost if the feature isn't used, otherwise I think i can make it cost 1 bit per entity to mark if there's a predicted entity too.

This is to fix https://github.com/lifescapegame/bevy_replicon/issues/67

### CODE REVIEW

**Comment by RJ** [comment]
If we made a new shared config struct that the client and server plugins take, this kind of feature can be opt in depending on the style of game.

how about something like:

```rust
struct RepliconConfiguration {
    // Makes the `PredictionTracker` available on the server, and supports client-predicted entity mapping.
    pub client_predictions: bool,
}
```

Another option might be to not insert the `PredictionTracker` resource by default, and make the user insert the resource on both client and server if they want to enable the feature (even though the client doesn't need it)

**Comment by RJ** [comment]
reverted the entity encoding commit and added a currently-failing test showing the edge case @Shatur pointed out. 

> "include additional array with entity to entity mappings instead. This array should come first and will be processed before replication and it just updates mappings on client. So additional callback won't be required. "

also pointed out there's no need for the prediction hit callback, since successful prediction means the predicted entity will receive (at least) the `Replication` component, so hits can be detected with a `Query<Entity, <With<Prediction>, Added<Replication>>>`

i will change it to using an entity mapping array as discussed.

**Review by Shatur — APPROVED** [review]

File: src/client.rs [hunk 1]
```diff
@@ -138,6 +143,33 @@ fn deserialize_tick(
...
+) -> Result<(), bincode::Error> {
+    let array_len: u16 = bincode::deserialize_from(&mut *cursor)?;
+    for _ in 0..array_len {
+        let server_entity = deserialize_entity(cursor)?;
+        let client_entity = deserialize_entity(cursor)?;
+
+        if let Some(entry) = entity_map.to_client().get(&server_entity) {
+            error!("received mapping from {server_entity:?} to {client_entity:?}, but already mapped to {entry:?}");
+            continue;
+        }
```
[the remark this answers is not in the document]
Comment (Shatur) [reply src/client.rs:159]: It actually a client code, maybe we should `panic!`?
This should never happen. If it does - the server is programmed wrong.
We should never panic on server, but panicking on client is fine. It could happen even if you misuse bevy.
